ISPN-14213: Add a statement why we provide no performance numbers (#10380)
infinispan/infinispan: documentation/src/main/asciidoc/topics/con_performance_metric_considerations.adoc

> :_content-type: CONCEPT

Reviewer: I don't think this is necessary if the file name has the `con_` prefix.

Author: Removed.

[id='performance-metric-considerations_{context}']
= Performance metric considerations

{brandname} includes so many configurable combinations that determining a single formula for performance metrics that covers all use cases is not possible.  

Additionally, consider the following inter-related factors that apply to {brandname}:

* Available CPU and memory resources in cloud environments
* Caches used in parallel
* Get, put, query balancing 
* Peak load and throughput limitations
* Querying limitations with data set
* Number of entries per cache
* Size of cache entries

> Given the number of different combinations and unknown external factors, providing a performance calculation that meets all {brandname} use cases is not possible, because you cannot compare one performance test to another test if any of the previously listed factors are different. 

Reviewer: Sentence is way too long.

Author: Split into two sentences.

> The purpose of this document is to provide details about use cases and architectures that can help you determine requirements for your {brandname} deployment.

Reviewer: I would move this sentence to the top. Maybe the opening sentence.

Author: Moved to start of module.

> You can run a basic performance test from CLI that collects limited performance metrics. You can customize the performance test by parameters, so that the test outputs results that might meet your needs. The results from the test can provide you with baseline metrics that you can then use to determine settings and resources for your {brandname} caching requirements. 

Reviewer: "from the {brandname} command line interface (CLI)"

Reviewer: Also please take a look at the one sentence per line recommendation. This is something the Infinispan doc set adheres to. 

https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

Author: Hi @oraNod . This recommendation is based on a document in draft, and the style does not seem to be evident on other projects I've worked on. I do find sticking to this recommendation difficult, because I don't know the "why" of it? The principle has not been explained to me very well. Is it recommended because it makes scanning an _.adoc_ file easier?

Reviewer: The fact that doc is in some kind of draft state doesn't really matter. The advantages that doc lists hold true and explain the why.

In addition to all that, one sentence per line helps with PR reviews. But, generally, it also makes a lot of sense because of [how asciidoc creates paragraphs](https://docs.asciidoctor.org/asciidoc/latest/blocks/paragraphs/).

You can find this in the Infinispan contributor's guide here: https://infinispan.org/docs/stable/titles/contributing/contributing.html#documentation_guidelines_general

Author: Thanks for sharing, @oraNod . I really need to know "why", as I've been asked to implement things and being told they don't break anything, so why should they hold up the progress of a PR. 

I generally follow internal guidelines, so I can stay up-to-date with the "one voice". Who updates the Infinispan Contributor's guide? For example, "contrctions" have been relaxed. I generally don't use them, but they are no longer slapped with the non-recommended guidelines.

Reviewer: @dfitzmau little things like this can have larger, cumulative effects down the road. Part of why I brought it up is that I see using ventilated prose (one sentence per line) as a good writing habit. It's not just a style guide thing.

I picked it up from one of the core devs on this project and try to evangelize it a little. I'd suggest going back and reading Dan Allen's take on the advantages. He pretty much says it all.

I'm not sure what you mean by "one voice"?

"Who updates the Infinispan Contributor's guide?" - The Infinispan team update the Contributor's Guide.
"For example, "contrctions" have been relaxed. I generally don't use them, but they are no longer slapped with the non-recommended guidelines." - Again, not sure what you mean by this.

Author: I think the topic of "preference" is highly debatable. Chat channels are devoted to what someone thinks is right or wrong. I prefer consistency, so that one can remain agile among products.

"simplifying" processes was also mentioned at a very high-level. Having too many sources to reference will eventually lead to confusion, as they start to contradict one another. 

The contributor guide you linked mentions "Avoid contractions. Use "do not" instead of "don’t".",

[id='performance-metric-considerations_{context}']
= Performance metric considerations
{brandname} includes so many configurable combinations that determining a single formula for performance metrics that covers all use cases is not possible.  

> The purpose of the _{brandname} performance considerations and tuning guidelines_ document is to provide details about use cases and architectures that can help you determine requirements for your {brandname} deployment.

Reviewer 2: Hi! This is the community title but in the product it's different. This will require a follow up PR. Either you can reword it so you don't mention the title name or you can create an attribute for the title that will have different value in the community and in the product. I can help

Additionally, consider the following inter-related factors that apply to {brandname}:

* Available CPU and memory resources in cloud environments
* Caches used in parallel
* Get, put, query balancing 
* Peak load and throughput limitations
* Querying limitations with data set
* Number of entries per cache
* Size of cache entries

Given the number of different combinations and unknown external factors, providing a performance calculation that meets all {brandname} use cases is not possible. 
You cannot compare one performance test to another test if any of the previously listed factors are different. 

> You can run a basic performance test from CLI that collects limited performance metrics. 

Reviewer: ```suggestion
You can run basic performance tests with the {brandname} CLI that collects limited performance metrics. 
```

Author: :heavy_check_mark:

> You can customize the performance test by parameters, so that the test outputs results that might meet your needs. 

Reviewer: ```suggestion
You can customize the performance test so that the test outputs results that meet your needs. 
```

Author: :heavy_check_mark:

> The results from the test can provide you with baseline metrics that you can then use to determine settings and resources for your {brandname} caching requirements. 

Reviewer: ```suggestion
Test results provide baseline metrics that can help you determine settings and resources for your {brandname} caching requirements. 
```

Author: :heavy_check_mark:

Measure the performance of your current settings and check if they meet your requirements. 

> If your needs are not met, optimize the settings and then re-measure their performance.

Reviewer: Honestly I think we can drop these two lines.

Author: I'll leave these sentences in because the SME requested them. These sentences might be obvious, but we really want to enforce that the user takes responsibility for determining their performance metrics. We cannot provide a "one-shot" performance metric that spans all customer user cases. Providing users with as much baseline help might help their reach their desired metric determination goals.
